src/black/handle_ipynb_magics.py

```diff
@@ -0,0 +1,422 @@
+    if cell_magic_finder.cell_magic is None:
+        return src, replacements
+    mask = get_token(src, cell_magic_finder.cell_magic.header)
+    replacements.append(Replacement(mask=mask, src=cell_magic_finder.cell_magic.header))
+    return f"{mask}\n{cell_magic_finder.cell_magic.body}", replacements
+
+
+def replace_magics(src: str) -> Tuple[str, List[Replacement]]:
+    """Replace magics within body of cell.
+
+    Note that 'src' will already have been processed by IPython's
+    TransformerManager().transform_cell.
+
+    Example, this
+
+        get_ipython().run_line_magic('matplotlib', 'inline')
+        'foo'
+
+    becomes
+
+        "5e67db56d490fd39"
+        'foo'
+
+    The replacement, along with the transformed code, are returned.
+    """
+    replacements = []
+    magic_finder = MagicFinder()
+    magic_finder.visit(ast.parse(src))
+    new_srcs = []
+    for i, line in enumerate(src.splitlines(), start=1):
+        if i in magic_finder.magics:
+            offsets_and_magics = magic_finder.magics[i]
+            if len(offsets_and_magics) != 1:  # pragma: nocover
+                # defensive check
+                raise UnsupportedMagic
+            col_offset, magic = (
+                offsets_and_magics[0].col_offset,
+                offsets_and_magics[0].magic,
+            )
+            mask = get_token(src, magic)
+            replacements.append(Replacement(mask=mask, src=magic))
+            line = line[:col_offset] + mask
+        new_srcs.append(line)
+    return "\n".join(new_srcs), replacements
+
+
+def unmask_cell(src: str, replacements: List[Replacement]) -> str:
+    """Remove replacements from cell.
+
+    For example
+
+        "9b20"
+        foo = bar
+
+    becomes
+
+        %%time
+        foo = bar
+    """
+    for replacement in replacements:
+        src = src.replace(replacement.mask, replacement.src)
+    return src
+
+
+def _is_ipython_magic(node: ast.expr) -> TypeGuard[ast.Attribute]:
+    """Check if attribute is IPython magic.
+
+    Note that the source of the abstract syntax tree
+    will already have been processed by IPython's
+    TransformerManager().transform_cell.
+    """
+    return (
+        isinstance(node, ast.Attribute)
+        and isinstance(node.value, ast.Call)
+        and isinstance(node.value.func, ast.Name)
+        and node.value.func.id == "get_ipython"
+    )
+
+
+@dataclasses.dataclass(frozen=True)
+class CellMagic:
+    header: str
+    body: str
+
+
+@dataclasses.dataclass
+class CellMagicFinder(ast.NodeVisitor):
+    """Find cell magics.
+
+    Note that the source of the abstract syntax tree
+    will already have been processed by IPython's
+    TransformerManager().transform_cell.
+
+    For example,
+
+        %%time\nfoo()
+
+    would have been transformed to
+
+        get_ipython().run_cell_magic('time', '', 'foo()\\n')
+
+    and we look for instances of the latter.
+    """
+
+    cell_magic: Optional[CellMagic] = None
```

Comment:
> Same, let's use a dataclass

Done - is this what you meant @JelleZijlstra ? (sorry having taken some time to address that, I wasn't aware that dataclasses would inherit from other classes - there, that's another new thing I've learned from this PR 💚  )

Reply:
Annnddd it looks like this broke compatibility with mypyc:

```console
~/programming/oss/black on mypyc-support-pt2 [$?⇡] via Python v3.8.5 (bm-venv) took 1s445ms 
❯ black setup.py --check -v
Traceback (most recent call last):
  File "/home/ichard26/programming/oss/black/bm-venv/bin/black", line 5, in <module>
    from black import patched_main
  File "src/black/__init__.py", line 47, in <module>
    from black.files import find_project_root, find_pyproject_toml, parse_pyproject_toml
  File "/home/ichard26/programming/oss/black/bm-venv/lib/python3.8/site-packages/black/files.py", line 26, in <module>
    from black.handle_ipynb_magics import jupyter_dependencies_are_installed
  File "src/black/handle_ipynb_magics.py", line 316, in <module>
    class CellMagicFinder(ast.NodeVisitor):
AttributeError: attribute '__dict__' of 'type' objects is not writable
```

(it's broken on main too but I'm currently working on mypyc integration on the `mypyc-support-pt2` branch). I'll deal with this myself since I need to assess how badly this broke things. The good news is that this should be easy to work around (hopefully just remove the dataclass decorator) since I've seen and dealt with issues like this before.